### ISSUE TITLE
Libhermit's use of `halt` and the `panic_handler`

### ISSUE THREAD

**Issue by jounathaen** [hermitcore/libhermit-rs#45]
When panicing, RustyHermit loops `arch::processor::halt()` infinitely. Uhyve threatens `hlt` instructions by shutting down the vm with exit code 0. (https://github.com/hermitcore/uhyve/issues/7)

Related Issue: https://github.com/hermitcore/uhyve/issues/11

 RustyHermit should not use `arch::processor::halt` when running in a VM but use the proper shutdown command.
- It should probably keep using a hlt loop when running on real hardware so that the error message can be observed on e.g. the monitor.
- It could write to the shutdown port, so that uhyve shuts down before the hlt-loop. I don't know how real Hardware handles `out` instructions to a non-existing port.

**Comment by stlankes** [comment]
Doesn't the patch solve this issue?

**Comment by jschwe** [comment]
Is this issue completely solved? 

I'd like to link to the [Travis CI build](https://travis-ci.com/github/hermitcore/uhyve/jobs/334213154) for commit 4a2381ccc192b3d699fd4b54fb007e297bac9e2d. In release mode it seems a race condition occured that caused a Panic due to a BorrowError: `[0][!!!PANIC!!!] /home/travis/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/cell.rs:798: already mutably borrowed: BorrowError`
Given the output I assume that the BorrowError occured in the kernel or demo and not in uhyve. 
The thing that makes this related to this isssue is that uhyve doesn't exit after the panic. 

I've saved the log output here so it doesn't get lost if someone reruns the build.
[log.txt](https://github.com/hermitcore/uhyve/files/4633138/log.txt)

_Originally posted by @jschwe in https://github.com/hermitcore/uhyve/issues/7#issuecomment-629103723_

**Comment by stlankes** [comment]
Yes, I guess that an interrupt handler is trying to borrow something, which is already borrowed. I will check it

**Comment by jschwe** [comment]
The reason I'm mentioning this here is because uhyve doesn't exit after the Panic.

**Comment by stlankes** [comment]
Ah, you tested with the current patch?

**Comment by jschwe** [comment]
The Travis CI build was for https://github.com/hermitcore/uhyve/commit/4a2381ccc192b3d699fd4b54fb007e297bac9e2d and tested against https://github.com/hermitcore/rusty-hermit/commit/18a0235850a34c7021dde1413c40c163c5c153dc

Regarding uhyve this means that the Halt State is ignored since https://github.com/hermitcore/uhyve/pull/14 is already merged, so if the panic handler still enters the halt state, this would explain why uhyve doesn't exit.

**Comment by stlankes** [comment]
I hope that PR #51 helps to fix the issue.

**Comment by jschwe** [comment]
[The travis job](https://travis-ci.com/github/hermitcore/libhermit-rs/jobs/337449215) for https://github.com/hermitcore/libhermit-rs/commit/fc5d048a82bf6dea618b54845a06caadad705a4c failed and entered the idle loop. Before merging the branch [all tests passed](https://github.com/hermitcore/libhermit-rs/runs/692214841).
This indicates that the issue is not yet solved, or perhaps flaky. 
The relevant part from the log:
<details><summary>Backtrace</summary>
<p>
```
$ sudo -E sudo -u $USER -E bash -c "HERMIT_VERBOSE=1 HERMIT_CPUS=2 $HOME/.cargo/bin/uhyve target/x86_64-unknown-hermit/debug/rusty_demo"

[0][INFO] Welcome to HermitCore-rs 0.3.30

[0][INFO] Kernel starts at 0x200000

[0][INFO] BSS starts at 0x47d680

[0][INFO] TLS starts at 0x47b5c8 (size 304 Bytes)

[0][INFO] Total memory size: 64 MB

[0][INFO] A pure Rust application is running on top of HermitCore!

[0][INFO] Heap: size 50 MB, start address 0x600000

[0][INFO] Heap is located at 0x600000 -- 0x3800000 (0 Bytes unmapped)

[0][INFO] 

[0][INFO] ===================== PHYSICAL MEMORY FREE LIST ======================

[0][INFO] 0x00000003800000 - 0x00000004000000

[0][INFO] ======================================================================

[0][INFO] 

[0][INFO] 

[0][INFO] ================== KERNEL VIRTUAL MEMORY FREE LIST ===================

[0][INFO] 0x00000003800000 - 0x00800000000000

[0][INFO] ======================================================================

[0][INFO] 

[0][INFO] 

[0][INFO] ========================== CPU INFORMATION ===========================

[0][INFO] Model:                   uhyve - unikernel hypervisor

[0][INFO] Frequency:               2800 MHz (from Hypervisor)

[0][INFO] SpeedStep Technology:    Not Available

[0][INFO] Features:                MMX SSE SSE2 SSE3 SSSE3 SSE4.1 SSE4.2 AVX AESNI RDRAND FMA MOVBE MCE FXSR XSAVE RDTSCP CLFLUSH TSC-DEADLINE X2APIC AVX2 AVX512F AVX512DQ AVX512CD AVX512BW AVX512VL BMI1 BMI2 RTM HLE FSGSBASE 

[0][INFO] Physical Address Width:  46 bits

[0][INFO] Linear Address Width:    48 bits

[0][INFO] Supports 1GiB Pages:     No

[0][INFO] ======================================================================

[0][INFO] 

[0][INFO] HermitCore-rs booted on 2020-05-20 at 09:54:51

[0][INFO] 

[0][INFO] ======================== PCI BUS INFORMATION =========================

[0][INFO] 00:00 Ethernet controller [0200]: Red Hat, Inc. Virtio network device [1AF4:1000], MemoryBar: 0xc000 (size 0x10)

[0][INFO] ======================================================================

[0][INFO] 

[0][INFO] IOAPIC v17 has 24 entries

[0][INFO] Disable IOAPIC timer

[1][INFO] Entering idle loop for application processor

[0][ERROR] Invalid Opcode (#UD) Exception: ExceptionStackFrame {

    instruction_pointer: 0x1ff56a,

    code_segment: 0x8,

    cpu_flags: 0x10206,

    stack_pointer: 0x1feea0,

    stack_segment: 0x10,

}

No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.

Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received

The build has been terminated
```


</p>
</details>

**Comment by jounathaen** [comment]
Further locations of `spin_loop_hint` which have to be replaced by `panic!`:

- src/arch/x86_64/kernel/processor.rs:795
- src/mm/mod.rs:114

**Comment by jounathaen** [comment]
Closed by 7decb24f
